### PR TITLE
TPC: Update CheckOfTrendings

### PR DESCRIPTION
**Pull request by mlesch**
Changes: new checks (range check and zero check), new visualization (e.g. running mean) and inclusion of uncertainties of data points in the calculation of mean

### CODE REVIEW

File: Modules/TPC/src/CheckOfTrendings.cxx [hunk 1]
```diff
@@ -17,188 +17,580 @@
...
-  Quality resultExpectedPhysicsValue = Quality::Null;
+  std::vector<Quality> qualities;
 
   auto mo = moMap->begin()->second;
+  if (!mo) {
+    ILOG(Fatal, Support) << "Monitoring object not found" << ENDM;
+  }
   auto* canv = dynamic_cast<TCanvas*>(mo->getObject());
-  TGraph* g = (TGraph*)canv->GetListOfPrimitives()->FindObject("Graph");
+  ;
```
Comment (wiechula) [inline Modules/TPC/src/CheckOfTrendings.cxx:126]:
```suggestion
```

File: Modules/TPC/src/CheckOfTrendings.cxx [hunk 2]
```diff
@@ -17,188 +17,580 @@
...
+  const double* yValues = g->GetY();
+  const double* yErrors = g->GetEY(); // returns nullptr for TGraph (no errors)
 
-  // If only one data point available, don't check quality for mean
-  if (NBins > 1) {
-    if (mCheckChoice == CheckChoiceMean || mCheckChoice == CheckChoiceBoth) {
+  bool UseErrors = true;
+  if (yErrors == nullptr) {
+    UseErrors = false;
+    ILOG(Error, Support) << "NO ERRORS" << ENDM;
```
Comment (wiechula) [inline Modules/TPC/src/CheckOfTrendings.cxx:143]: I guess this can be warning or info

File: Modules/TPC/src/CheckOfTrendings.cxx [hunk 3]
```diff
@@ -17,188 +17,580 @@
...
+  if (mZeroCheck) {
+    checkMessage = "";
+    checkResult = Quality::Null;
+    int pointNumber = mPointToTakeForZeroCheck;
+    if (NBins < pointNumber) {
+      pointNumber = NBins;
+    }
+
+    const std::vector<double> v(yValues + NBins - pointNumber, yValues + NBins);
+    bool AllZeros = std::all_of(v.begin(), v.end(), [](double i) { return std::abs(i) == 0.0; });
```
Comment (wiechula) [inline Modules/TPC/src/CheckOfTrendings.cxx:275]:
```suggestion
    const bool AllZeros = std::all_of(v.begin(), v.end(), [](double i) { return std::abs(i) == 0.0; });
```

File: Modules/TPC/src/CheckOfTrendings.cxx [hunk 4]
```diff
@@ -17,188 +17,580 @@
...
+    checks.push_back(checkMessage);
+  }
+
+  // Quality aggregation
+  if (qualities.size() >= 1) {
+    auto Worst_Quality = std::max_element(qualities.begin(), qualities.end(),
+                                          [](const Quality& q1, const Quality& q2) {
+                                            return q1.isBetterThan(q2);
+                                          });
+    int Worst_Quality_Index = std::distance(qualities.begin(), Worst_Quality);
```
Comment (wiechula) [inline Modules/TPC/src/CheckOfTrendings.cxx:293]: this variable is not used

File: Modules/TPC/src/CheckOfTrendings.cxx [hunk 5]
```diff
@@ -17,188 +17,580 @@
...
 }
 
 std::string CheckOfTrendings::getAcceptedType() { return "TCanvas"; }
 
 void CheckOfTrendings::beautify(std::shared_ptr<MonitorObject> mo, Quality checkResult)
 {
-  auto* c1 = dynamic_cast<TCanvas*>(mo->getObject());
-  TGraph* h = (TGraph*)c1->GetListOfPrimitives()->FindObject("Graph");
-  TPaveText* msg = new TPaveText(0.7, 0.85, 0.9, 0.9, "NDC");
+  auto* canv = dynamic_cast<TCanvas*>(mo->getObject());
```
Comment (wiechula) [inline Modules/TPC/src/CheckOfTrendings.cxx:329]: `nullptr` check needed?

File: Modules/TPC/src/CheckOfTrendings.cxx [hunk 6]
```diff
@@ -17,188 +17,580 @@
...
+    TGraph* MeanGraph = new TGraph();
+    TGraph* StddevGraphMediumUp = new TGraph();
+    TGraph* StddevGraphMediumDown = new TGraph();
+    TGraph* StddevGraphBadUp = new TGraph();
+    TGraph* StddevGraphBadDown = new TGraph();
+
+    const int nPoints = h->GetN();
+    const double* yValues = h->GetY();
+    const double* yErrors = h->GetEY(); // returns nullptr for TGraph (no errors)
+    bool UseErrors = true;
```
Comment (wiechula) [inline Modules/TPC/src/CheckOfTrendings.cxx:407]: Naming convention!

**Review by wiechula — CHANGES_REQUESTED** [review]
Please carefully check the naming convention of functions and variables.

**Comment by mlesch** [comment]
@wiechula thanks for the review! I implemented all requested changes

File: Modules/TPC/include/TPC/CheckOfTrendings.h [hunk 7]
```diff
@@ -17,6 +17,8 @@
 #define QC_MODULE_TPC_CHECKOFTRENDINGS_H
 
 #include "QualityControl/CheckInterface.h"
+#include <TGraph.h>
+#include <TCanvas.h>
```
Comment (wiechula) [inline Modules/TPC/include/TPC/CheckOfTrendings.h:21]: Forward declare:

```suggestion
class TGraph;
class TCanvas;
```

And move
```
#include <TGraph.h>
#include <TCanvas.h>
```
to the `cxx`

**Review by wiechula — CHANGES_REQUESTED** [review]

**Comment by mlesch** [comment]
@wiechula Thanks, I changed it to forward declaration

**Review by wiechula — APPROVED** [review]
Looks fine I think. Do you want to remove the WIP?

**Pull request merged**